### PR TITLE
feat(runtime/debug): one-command replay of build_output via debug/run.py

### LINKED ISSUE #1327: [Feature] Make build_output runnable through simpler without manual SceneTestCase wrapping
### Summary

When debugging a kernel produced by `ir.compile`, I sometimes want to skip pypto's Python front-end entirely and re-run the existing `build_output/<jit_dir>/` contents (kernel cpp/so + orchestration cpp/so + `kernel_config.py`) directly through simpler's `SceneTestCase` / `scene_test` machinery — for example to bisect a runtime stall against a different simpler revision, or to iterate on the orchestration cpp by hand-editing it.

Today this requires a substantial rewrite of the auto-generated `kernel_config.py`. The generated file only contains `KERNELS` (list), `ORCHESTRATION` (dict), and `RUNTIME_CONFIG` — none of which simpler can consume directly. To run via simpler I had to:

1. Add `from simpler.task_interface import ArgDirection as D` and re-derive every kernel's `signature` (the `D.IN` / `D.OUT` / `D.INOUT` list) by hand-reading the orchestration cpp (`add_input` / `add_output` / `add_inout` calls per `params_t<i>`).
2. Re-derive the orchestration's external-tensor signature from the `from_tensor_arg(orch_args.tensor(N))` lines and which of those are written via `add_output` somewhere.
3. Convert the flat `KERNELS` + `ORCHESTRATION` dicts into simpler's nested `CALLABLE = {\"orchestration\": {..., \"signature\": [...]}, \"incores\": [{..., \"signature\": [...]}, ...]}` shape.
4. Wrap the result in a `@scene_test(level=2, runtime=...)` `SceneTestCase` subclass, write `CASES`, and implement `generate_args` + `compute_golden` (importing `build_tensor_specs` / golden helpers from the original pypto-lib source module via a `sys.path` hack into the repo root, since `build_output/<jit_dir>/` isn't on the import path).
5. Add `if __name__ == \"__main__\": SceneTestCase.run_module(__name__)`.

Steps 1–3 are pure mechanical translation of information pypto already has at compile time. The only step that genuinely needs user input is the test inputs / golden — i.e. step 4's `generate_args` + `compute_golden`.

### Motivation / Use Case

Concrete debugging scenario that triggered this:

I have a `build_output/_jit_moe_expert_test_<ts>/` from `models/deepseek/v4/deepseek_v4_decode_moe_expert.py`. The dispatch ran into a runtime stall (10 AIVs busy, 0 AICs, scheduler `in_flight=0`, `Thread N: PTO2 timeout after 800001 idle iterations`). To narrow it down I want to:

- Re-run the same build_output against a different simpler / pto-isa pin without re-running pypto's IR compile (which is slow and would invalidate the artifact under test).
- Edit `orchestration/moe_expert_test.cpp` by hand and re-run, since the suspected root cause is a missing ringbuffer notify that originates from codegen — re-emitting via `pl.jit` would just regenerate the buggy version.
- Run via `pytest kernel_config.py -p a2a3` so I get the full `SceneTestCase` infra (per-case dispatch, swimlane, PMU, `--dump-tensor`, etc.).

None of this requires any IR-level work. It just requires that the generated artifact be re-runnable. Today the boilerplate above blocks that.

### Proposed API / Behavior

Two reasonable directions; I'd be happy with either, but my preference is **Option A** because it puts the metadata where it semantically belongs.

**Option A — Enrich `kernel_config.py` codegen so simpler can consume it directly.**

At `compile_program` time, pypto already knows each task's `add_input` / `add_output` / `add_inout` direction (it's emitting the orchestration cpp from the IR). Emit that as a `signature` field on each `KERNELS` entry and on `ORCHESTRATION`, and additionally emit a sibling `CALLABLE` dict in the simpler-expected shape:

```python
# kernel_config.py (additions in **bold**)
ORCHESTRATION = {
    \"source\": ...,
    \"function_name\": \"aicpu_orchestration_entry\",
    **\"signature\": [D.IN, D.IN, ..., D.OUT, D.OUT],**  # 18 entries
}
KERNELS = [
    {\"func_id\": 0, \"name\": \"x_local_q\", ..., **\"signature\": [D.IN, D.OUT, D.OUT]**},
    ...
]
**CALLABLE = {\"orchestration\": ORCHESTRATION, \"incores\": KERNELS}**  # alias for simpler
```

With that, the user-side wrapper collapses to a thin `SceneTestCase` whose `CALLABLE = ...` line just imports from `kernel_config`. Steps 1–3 above disappear; only the test-input / golden step remains, which is genuinely user-supplied.

(Optionally, the codegen could also drop a `kernel_config_scene_test.py` template alongside that contains the `@scene_test` class skeleton with TODO markers for `generate_args` / `compute_golden`. Not required, but nice.)

**Option B — Add a `pypto.runtime.execute_compiled` debug entrypoint that takes simpler-style input.**

Add a `replay_compiled(work_dir, case_name=..., platform=..., device_id=..., simpler_kwargs=...)` (name TBD) that:

- Loads `kernel_config.py`.
- Skips `KernelCompiler` / `_patch_orchestration_headers` / pto-isa cloning when the user passes `--skip-compile` (i.e. trust the existing `.so` files under `kernels/<core>/` and `orchestration/`).
- Builds the simpler `ChipCallable` directly from those `.so`'s + signatures (signatures still need to be present, so this depends on the same codegen change as Option A, just hidden behind a different surface).
- Lets the user pass torch tensors / scalars for the orch args, the same way `execute_compiled` does today, but routes through simpler's runtime test harness so swimlane / PMU / `--dump-tensor` / `-c` pinning all work.

Option B is strictly more work because it duplicates simpler's `SceneTestCase.run_module` CLI; that's why I prefer Option A.

### Alternatives Considered

- **Hand-write the `SceneTestCase` once and check it in.** That's what I did this session. It works but duplicates information already in the orchestration cpp, drifts as soon as the kernel signature changes, and has to be redone for every new build_output.
- **Re-run via `python <kernel>.py -p ...` (the normal flow).** Doesn't solve the use case — that re-runs pypto's IR compile, which regenerates the artifact under test and so cannot reproduce a bug specific to a particular `build_output/<jit_dir>/`.
- **Manually invoke `pypto.runtime.execute_compiled(work_dir, args, ...)` from a small driver script.** Works for one-shot reproduction but loses the `SceneTestCase` infra (multi-case dispatch, swimlane, PMU, `-c` pin, dump-tensor, parallel scheduler).

### Additional Context

- For reference, the wrapper I had to write by hand for the moe_expert build_output is ~150 lines. About 100 of those lines (signatures + CALLABLE shape) are pure mechanical translation that pypto could emit; the remaining ~50 (CASES + generate_args + compute_golden) genuinely need user input.
- Git Commit ID: `4b69a87696a3528dfc50d4cfbeccbe98574378e1`
- Host Platform: Linux (aarch64)
- NPU Kind: N/A (not hardware-specific — this is a codegen / API ergonomics issue)

### PR DESCRIPTION
**Pull request by luohuan19**
## Summary

Closes #1327 via Option B — a dedicated `pypto.runtime.debug` replay path,
plus an auto-emitted `debug/run.py` so that re-executing any
`build_output/<jit_dir>/` is always **one command**:

```bash
python build_output/<jit_dir>/debug/run.py
```

No more hand-rolling a `SceneTestCase` wrapper, re-deriving `signature`
lists from the orchestration cpp, or re-running the slow IR-compile path
just to reproduce a runtime bug pinned to a specific build_output.

### What's in this PR

- **`pypto.runtime.debug.replay`** — re-execute an existing build_output
  directly from its on-disk `kernel_config.py`. Supports `--validate`
  against a sibling `golden.py`, `--log-level` for runtime verbosity,
  and skips the front-end IR compile entirely (the bug-under-test
  artefact is never regenerated).
- **`golden.py` generation** — `ir.compile()` now writes a self-contained
  `golden.py` next to the build_output with `generate_inputs()` +
  `compute_golden()`. Callable globals referenced by the source program
  are inlined so the generated file is import-clean.
- **`debug/run.py` auto-emission** — `CompiledProgram` /
  `DistributedCompiledProgram` write a thin runner that calls
  `replay`, materialises inputs (from `golden.generate_inputs()` when
  available, else from shape/dtype metadata for the JIT path), and
  validates against `compute_golden`. JIT path additionally exposes a
  `_user_compare(<param_names>)` hook for hand-rolled
  `torch.allclose` assertions.
- **CLI lifted into `replay._main`** — the emitted `debug/run.py` is now
  a thin shim that forwards `sys.argv` to `replay._main` with three
  keyword hooks (`inline_inputs`, `user_compare`, `default_platform`).
  Adding a new replay flag no longer requires a parallel edit in
  `run_script_writer`; `--validate` is `BooleanOptionalAction` so
  `--no-validate` is discoverable via `--help`. A drift test asserts
  the generated shim cannot silently re-grow an argparse block.
- **Edit-`.pto`-and-replay** — when `ptoas/<unit>.pto` is newer than
  its sibling `.cpp`, replay reruns `ptoas` and splices the
  preprocessed body into matching `kernels/<core>/<func>.cpp` between
  `BEGIN`/`END` sentinels, then rebuilds the `.so`. Sentinel format
  and the preprocess pipeline are guarded by drift tests against
  `pto_backend`. Opt out via `--no-rebuild-from-pto` or
  `PYPTO_REBUILD_FROM_PTO=0`.
- Toggle: `PYPTO_EMIT_DEBUG_RUNNER=0` disables emission globally for
  bulk-compile pipelines (the `replay` CLI is still usable directly).
- Docs (`docs/{en,zh-cn}/dev/03-runtime-dfx.md`) describe the workflow.
- Tests in `tests/ut/runtime/` and `tests/ut/backend/`.

### Relationship to #1327

The issue suggested two options. This PR ships **Option B** (a runtime
replay entrypoint) because it does not require freezing a new contract
into `kernel_config.py` codegen — the `signature` / `CALLABLE` shape
proposed in Option A would have been a heavier external-API change.
The same end-user pain point is solved: re-running an existing
build_output no longer needs manual wrapping.

## Testing

- [x] `tests/ut/runtime/test_run_script_writer.py` — emitted script
      shape, env-var gating, `_user_compare` placement on JIT path only,
      drift guard against argparse/RunConfig re-appearing in the shim
- [x] `tests/ut/runtime/test_replay.py` — golden / replay coverage,
      `_main` keyword-hook contracts (inline fallback, user_compare vs
      golden validation, default_platform baking)
- [x] `tests/ut/runtime/test_pto_rebuild.py` — pto->cpp splice path,
      staleness detection, sentinel matching
- [x] `tests/ut/backend/test_ptoas_sentinels_match.py` — drift guard
      tying sentinel format to `pto_backend`
- [x] Doc updates kept in lockstep (en + zh)

## Related Issues

Closes #1327

### CODE REVIEW

File: python/pypto/runtime/debug/replay.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+    """Compute expected outputs via ``golden.py`` and compare against actuals.
+
+    Actual outputs (already written in place by orchestration) are matched
+    by name against the ``__outputs__`` list. Expected outputs are produced
+    by cloning the actual output tensors (so dtype/shape match) and letting
+    ``compute_golden`` populate them from the user inputs. Comparison uses
+    :func:`torch.testing.assert_close` with the tolerances declared on the
+    ``golden.py`` module (``RTOL`` / ``ATOL``, defaulting to ``1e-5``).
+    """
+    spec = importlib.util.spec_from_file_location(
```
[the remark this answers is not in the document]
Comment (luohuan19) [reply python/pypto/runtime/debug/replay.py]: Fixed in `e2163228` — `replay()` now loads `golden.py` once and shares the module between input loading and output validation.

File: python/pypto/runtime/debug/replay.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    )
+    named_inputs = _load_named_inputs_from_golden(args.work_dir)
+    tensors = [v for _, v in named_inputs]
+    replay(
+        args.work_dir,
+        *tensors,
+        config=config,
+        recompile=not args.no_recompile,
+        validate=args.validate,
+    )
```
[the remark this answers is not in the document]
Comment (luohuan19) [reply python/pypto/runtime/debug/replay.py]: Partially addressed in `e2163228`: the duplicate load inside `replay(validate=True)` is gone (single load via `_load_golden_module`). The CLI still pre-loads inputs once before calling `replay` because the positional-tensor signature requires the tensor list up front; total load count is 2 (down from 3).

File: python/pypto/runtime/debug/__init__.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+"""PyPTO runtime debug helpers.
+
+Not for production use. Modules under ``pypto.runtime.debug`` are intended
+for offline debugging of compiled artefacts (re-execution, tracing, etc.).
+"""
+
+from .replay import invalidate_binary_cache, replay
+from .run_script_writer import write_run_script
+
+__all__ = ["replay", "invalidate_binary_cache", "write_run_script"]
```
[the remark this answers is not in the document]
Comment (luohuan19) [reply python/pypto/runtime/debug/__init__.py]: Fixed in `e2163228` — `__all__` sorted alphabetically.

File: python/pypto/runtime/debug/replay.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+    named_tensors: list[tuple[str, torch.Tensor]] | None = None
+    if validate:
+        named_defaults = _load_named_inputs_from_golden(work_dir)
+        if len(tensors) != len(named_defaults):
+            raise ValueError(
+                f"replay(validate=True): expected {len(named_defaults)} tensors "
+                f"(orchestration parameter count from {work_dir}/golden.py), "
+                f"got {len(tensors)}"
+            )
+        named_tensors = [(n, t) for (n, _), t in zip(named_defaults, tensors, strict=True)]
```
[the remark this answers is not in the document]
Comment (luohuan19) [reply python/pypto/runtime/debug/replay.py]: Fixed in `e2163228` — `replay(validate=True)` now raises a clear `TypeError` naming the parameter when a non-`torch.Tensor` is passed, instead of failing later at `.cpu()`.

File: python/pypto/runtime/debug/run_script_writer.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+) -> Path:
+    """Emit ``<work_dir>/debug/run.py``.
+
+    Args:
+        work_dir: A ``build_output/<jit_dir>/`` produced by ``ir.compile()``.
+        param_infos: Orchestration parameter metadata, in call order.
+            Obtain via :func:`pypto.ir.compiled_program.extract_param_infos`.
+        platform: Default platform baked into the generated CLI's
+            ``--platform`` argument. ``None`` falls back to ``"a2a3sim"``.
+
```
[the remark this answers is not in the document]
Comment (luohuan19) [reply python/pypto/runtime/debug/run_script_writer.py:67]: Fixed in `e2163228` — `_render()` now uses `a2a3sim` to match the docstring.

File: python/pypto/runtime/debug/run_script_writer.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    torch_dtype = _to_torch_dtype(p.dtype)
+    dtype_str = _TORCH_DTYPE_NAMES.get(torch_dtype, "torch.float32") if torch_dtype else "torch.float32"
+    concrete_shape = tuple(d if d >= 0 else 1 for d in p.shape)
+    has_dynamic = any(d < 0 for d in p.shape)
+    comment = "  # NOTE: dynamic dim(s) filled with 1 — edit as needed" if has_dynamic else ""
+
+    if p.direction == ParamDirection.Out:
+        return f"{p.name} = torch.zeros({concrete_shape!r}, dtype={dtype_str}){comment}"
+    return f"{p.name} = torch.randn({concrete_shape!r}, dtype={dtype_str}){comment}"
```
[the remark this answers is not in the document]
Comment (luohuan19) [reply python/pypto/runtime/debug/run_script_writer.py:207]: Fixed in `e2163228` — `_init_expr_for` now branches on dtype family: `torch.bool` → `torch.randint(0, 2, ...)`, integer dtypes → `torch.randint(0, 8, ...)`, float dtypes keep `torch.randn`.

File: tests/ut/runtime/test_replay.py [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+    bin_file = _touch(work_dir / "cache" / "incore_aiv_foo.bin")
+    so_file = _touch(work_dir / "kernels" / "aiv" / "foo.so")
+    with patch.object(replay_module, "execute_compiled"):
+        replay(work_dir, recompile=False)
+    assert bin_file.exists()
+    assert so_file.exists()
+
+
+def test_replay_missing_kernel_config_raises(tmp_path: Path) -> None:
+    with pytest.raises(FileNotFoundError, match="kernel_config.py"):
```
[the remark this answers is not in the document]
Comment (luohuan19) [reply tests/ut/runtime/test_replay.py]: Fixed in `e2163228` — three `pytest.raises(..., match=...)` patterns at lines 140/197/239 now use raw-string regex with escaped dots.

**Review by lyfne123 — APPROVED** [review]
